rfd/0157-aws-eks-discover.md

```diff
@@ -0,0 +1,205 @@
+message UIDiscoverEKSClusterEnrollEvent {
+  DiscoverMetadata metadata = 1;
+  DiscoverResourceMetadata resource = 2;
+  DiscoverStepStatus status = 3;
+  int64 selected_resources_count = 4;
+}
+```
+
+## Security
+
+New permissions will be required for the AWS OIDC integration to perform the necessary tasks. It will need 
+to be able to list EKS clusters as well as associate an OIDC provider with them. An associated OIDC provider 
+can generate authentication tokens that give full control over the cluster, so a careful approach to their 
+security should be taken.
+
+Tokens will be signed by a dedicated OIDC Certificate Authority (which already exists and is used for 
+AWS integration communications), ensuring that tokens don't overlap with regular CAs access by the users, 
+like the User CA or JWTSigner CA. The audience field value `kubernetes` will also be unique to this type of token. 
+And also user/group claim field names, such as `teleport-eks-oidc-groups`, will specify them as used for this purpose only.
+
+End users will never have access to the process of generating these tokens; we will use them only to perform the 
+initial installation of the `teleport-kube-agent` Helm chart. The agent itself will be configured with the standard 
+service account permissions it requires.
+
+After the Helm chart installation succeeds, Teleport will no longer need to have an OIDC provider associated with 
+the enrolled EKS cluster. Clients can then remove the association or even impose a block on OIDC provider 
+associations for the cluster if they wish.
+
+## Future considerations
+
+At the moment, the association of an OIDC provider with an EKS cluster allows for the best automation regarding cluster
+enrollment. This approach completely sidesteps the IAM method of EKS authentication, which currently lacks the flexibility
+we would require. Currently, IAM authentication in EKS is controlled by the `aws-auth` config map, which needs to be
+manually edited to allow IAM roles to have access to the cluster, and there's no API for that. Amazon is working on
+implementing an API for controlling `aws-auth`, but it's still work in progress (https://github.com/aws/containers-roadmap/issues/185).
+Once the aforementioned API is implemented, we might review the access pattern for the AWS EKS integration and compare the OIDC provider
+solution with the new API.
+
+## Alternative
+
+We could enroll clusters not by installing the kube agent through the Helm chart, but by proxying requests through the
+Kube service running outside of the EKS cluster. However, this would require clients to run a dedicated Kube service
+and would make permissions management more complicated due to the aforementioned issue with the EKS `aws-auth` config map.
+We would also lose the capability of auto-discovering apps inside EKS clusters, since the current Kubernetes App Auto-Discovery
+feature works only with the Teleport agent deployed inside the cluster.
```

Comment:
For the DiscoveryService we ended up deploying it alongside Auth for Cloud Tenants.
This removed the requirement for the user to have to run/manage yet another process, easing the process.

I was wondering if we could also start a Kube Service in Cloud and use that to proxy requests.
Is the Kube Apps Auto-Discovery support not possible at all, or would it require more development?
I think it would be really cool to remove that burden from them

Reply:
Kube App Discovery currently only works when agent is within the cluster. It's possible to improve on it, but that's additional work. 
As it's currently proposed in the RFD, there's no burden of starting Kube service as we enrol clusters just by installing kube agent into the cluster.

Reply:
Got it, thank you.
Totally agreed :+1:

Reply:
>As it's currently proposed in the RFD, there's no burden of starting Kube service as we enrol clusters just by installing kube agent into the cluster.

Assuming the discovered cluster has nodes :smile: